---
title: "Differential expression analysis of nAChR subunits in _Apis mellifera_"
author: "Federico Lopez, Alicja Witwicka"
date: '`r Sys.Date()`'
output:
  github_document:
    toc: yes
  pdf_document: 
    fig_caption: yes
    toc: yes
  html_document:
    toc: yes
number_sections: true

---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  include = FALSE,
  out.height = "\textheight",
  out.width = "\textwidth"
)
```

# Load libraries
```{r}
load_pkgs <- function(pkg) {
  sapply(pkg, require, character.only = TRUE)
}

cran_pkgs <- c(
  "BiocManager", "tidyverse", "RColorBrewer", "pheatmap", "styler", "here"
)
load_pkgs(cran_pkgs)

bioconductor_pkgs <- c(
  "biomaRt", "rhdf5", "tximport", "DESeq2", "limma", "RUVSeq"
)
load_pkgs(bioconductor_pkgs)
```

# Set a custom ggplot theme
```{r}
# Set a custom ggplot theme
# devtools::install_github("cttobin/ggthemr")
library(ggthemr)
custom_palette <- c(
  "#9E9E9E", "#59BDEF", "#EBCC2A", "#E1AF00", "#F27C8D", "#7B73D1", "#7B9FE0",
  "#F5CC7F", "#66C2A5", "#28A7B6", "#F2CB57", "#F2A057", "#F2845C"
)
# ggthemr::colour_plot(custom_palette)

custom_theme <- ggthemr::define_palette(
  swatch = custom_palette,
  gradient = c(lower = "#59BDEF", upper = "#FF6B5A")
)
ggthemr::ggthemr(custom_theme)
```

# Create directory for results
```{r}
# here::here()
# Create directory for results
dir.create(here::here("results", "2020-07-28-gene_expression"),
  recursive = TRUE
)

dir.create(here::here(
  "results", "2020-07-28-gene_expression",
  "supplementary"
),
recursive = TRUE
)
```

# Read tables of samples and covariates
```{r}
# Jasper et al. 2015
# Set sample identifiers using the names of the kallisto output directories
jasper_ids <- dir(here::here(
  "2015-jasper_worker_tissues", "2020-06-16-kallisto", "results"
))

# Read metadata table
jasper_table <- read.csv(
  here::here(
    "amel_metadata",
    "2015-jasper_worker_tissues_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

jasper_files <- here::here(
  "2015-jasper_worker_tissues", "2020-06-16-kallisto", "results",
  jasper_table$sample,
  "abundance.h5"
)

names(jasper_files) <- jasper_table$sample

# Check that sample names match
all(jasper_ids %in% jasper_table$sample)

# Check that all files exist
all(file.exists(jasper_files))

# Select the worker "brain" samples
jasper_brain_table <- jasper_table %>%
  dplyr::filter(tissue == "brain")

jasper_brain_files <- jasper_files[grep("brain", jasper_files)]

# Manfredini et al. 2015
# Set sample identifiers using the names of the kallisto output directories
amel_manfredini_ids <- dir(here::here(
  "2015-manfredini_queen_brains", "2020-06-18-kallisto", "results"
))

# Read metadata table
amel_manfredini_table <- read.csv(
  here::here(
    "amel_metadata",
    "2015-manfredini_queen_brains_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

amel_manfredini_files <- here::here(
  "2015-manfredini_queen_brains", "2020-06-18-kallisto", "results",
  amel_manfredini_table$sample,
  "abundance.h5"
)

names(amel_manfredini_files) <- amel_manfredini_table$sample

# Check that sample names match
all(amel_manfredini_ids %in% amel_manfredini_table$sample)

# Check that all files exist
all(file.exists(amel_manfredini_files))

# Liberti et al. 2015
# Set sample identifiers using the names of the kallisto output directories
liberti_ids <- dir(here::here(
  "2015-liberti_queen_brains", "2020-06-18-kallisto", "results"
))

# Read metadata table
liberti_table <- read.csv(
  here::here(
    "amel_metadata",
    "2015-liberti_queen_brains_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

liberti_files <- here::here(
  "2015-liberti_queen_brains", "2020-06-18-kallisto", "results",
  liberti_table$sample,
  "abundance.h5"
)

names(liberti_files) <- liberti_table$sample

# Check that sample names match
all(liberti_ids %in% liberti_table$sample)

# Check that all files exist
all(file.exists(liberti_files))

# Christen et al. 2015
# Set sample identifiers using the names of the kallisto output directories
christen_ids <- dir(here::here(
  "2018-christen_worker_brains", "2020-06-18-kallisto", "results"
))

# Read metadata table
christen_table <- read.csv(
  here::here(
    "amel_metadata",
    "2018-christen_worker_brains_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

christen_files <- here::here(
  "2018-christen_worker_brains", "2020-06-18-kallisto", "results",
  christen_table$sample,
  "abundance.h5"
)

names(christen_files) <- christen_table$sample

# Check that sample names match
all(christen_ids %in% christen_table$sample)

# Check that all files exist
all(file.exists(christen_files))
```

# Combine input files and information about the samples
```{r}
# Castes
castes_files <- c(
  jasper_brain_files, amel_manfredini_files, liberti_files, christen_files
)
castes_table <- dplyr::bind_rows(
  jasper_brain_table, amel_manfredini_table, liberti_table, christen_table
) %>%
  dplyr::select(sample, caste, tissue, study, species)

# Tissues
tissues_files <- jasper_files
tissues_table <- jasper_table
```

# Associate transcripts to genes
```{r}
biomaRt::listMarts(host = "metazoa.ensembl.org")
# biomaRt::listMarts(host = "eg40-metazoa.ensembl.org") # version 40
# biomaRt::listEnsemblArchives()

amel_mart <- biomaRt::useMart(
  biomart = "metazoa_mart",
  dataset = "amellifera_eg_gene",
  host = "metazoa.ensembl.org" # version 47
)

amel_tx2gene <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
  mart = amel_mart
)
```

# Import `kallisto` quantifications using `tximport`
```{r}
# Castes
castes_txi <- tximport::tximport(
  castes_files,
  type = "kallisto",
  tx2gene = amel_tx2gene,
  txOut = FALSE
)

# Tissues
tissues_txi <- tximport::tximport(
  tissues_files,
  type = "kallisto",
  tx2gene = amel_tx2gene,
  txOut = FALSE
)
```

# Create vector of nAChR gene identifiers
```{r}
amel_nachrs <- c(
  "GB42850", "GB42644", "GB47845", "GB43275", "GB43064", "GB43416",
  "GB53053", "GB40923", "GB53427", "GB53055", "GB53428"
)
```

# Analysis of castes
## Identify hidden batch effects using `RUVseq`
```{r}
# Create a DESeqDataSet object
castes_dds <- DESeq2::DESeqDataSetFromTximport(castes_txi,
  colData = castes_table,
  design = ~caste
)

# Choose the reference level for caste factor
castes_dds$caste <- relevel(castes_dds$caste, ref = "worker")

# Pre-filter the low-count genes
dim(castes_dds)
keep <- rowSums(DESeq2::counts(castes_dds) >= 10) >= 3
castes_dds <- castes_dds[keep, ]
dim(castes_dds)

# Run the DESeq2 pipeline and extract results
castes_dds <- DESeq2::DESeq(castes_dds, test = "Wald", betaPrior = FALSE)
castes_res <- DESeq2::results(castes_dds)

# Identify hidden batch effects using RUVSeq
set <- EDASeq::newSeqExpressionSet(DESeq2::counts(castes_dds,
  normalized = FALSE
))
idx <- rowSums(DESeq2::counts(set) >= 10) >= 3
set <- set[idx, ]
set <- EDASeq::betweenLaneNormalization(set, which = "upper")
notsig <- rownames(castes_res)[which(castes_res$pvalue > 0.1)]
empirical <- rownames(set)[rownames(set) %in% notsig]
set <- RUVSeq::RUVg(set, empirical, k = 2)
covars <- cbind(set$W_1, set$W_2)
```

## Add factor(s) of unwanted variation to the design in `DESeqDataSet` object
```{r}
castes_ddsruv <- castes_dds
castes_ddsruv$W1 <- set$W_1
castes_ddsruv$W2 <- set$W_2
design(castes_ddsruv) <- ~ W1 + W2 + caste
```

## Assess the relationships among samples after removing batch effects
```{r}
# Apply the variance stabilizing transformation
# The transformed data are on the log2 scale
castes_vsdruv <- DESeq2::vst(castes_ddsruv, blind = FALSE)

# Remove batch effects from the variance-stabilized data
assay(castes_vsdruv) <- limma::removeBatchEffect(assay(castes_vsdruv),
  design = model.matrix(~caste, data = castes_table),
  covariates = covars
  # covariates = set$W_1
)

castes_pca_data <- DESeq2::plotPCA(castes_vsdruv,
  intgroup = c("study", "caste"),
  ntop = 500, returnData = TRUE
)

percent_var <- round(100 * attr(castes_pca_data, "percentVar"))

ggplot(castes_pca_data, aes(
  x = PC1, y = PC2,
  color = caste, shape = study
)) +
  geom_point(
    size = 4, stroke = 1
  ) +
  xlab(paste0("PC1 [", percent_var[1], "%]")) +
  ylab(paste0("PC2 [", percent_var[2], "%]")) +
  coord_fixed(ratio = 1) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "plain"),
    plot.subtitle = element_text(hjust = 0.5, size = 14, face = "plain"),
    legend.title = element_blank(),
    legend.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text = element_text(size = 12)
  ) +
  scale_shape_manual(values = 1:12) +
  labs(
    title = "Relationships among samples from brains of honey bees",
    subtitle = "(batch effects removed)"
  )

ggsave(
  filename = here::here(
    "results", "2020-07-28-gene_expression", "supplementary",
    "amel_castes_vst_counts_batch_effects_removed_pca.pdf"
  ),
  width = 8,
  height = 6,
  units = "in"
)
```

## Differential gene expression analysis including batch factors
```{r}
# Run DESeq2 pipeline
castes_ddsruv <- DESeq2::DESeq(castes_ddsruv, test = "Wald", betaPrior = FALSE)
DESeq2::resultsNames(castes_ddsruv)

# Extract results
castes_resruv <- DESeq2::results(
  castes_ddsruv,
  contrast = c("caste", "queen", "worker"),
  lfcThreshold = 0,
  alpha = 0.1,
  pAdjustMethod = "BH"
)

DESeq2::summary(castes_resruv)
```

## Create table of `DESeq2` Wald test results for castes
```{r}
castes_resruv_df <- castes_resruv %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% amel_nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "GB42850" ~ "alpha1",
    ens_gene == "GB42644" ~ "alpha2",
    ens_gene == "GB47845" ~ "alpha3",
    ens_gene == "GB43275" ~ "alpha4",
    ens_gene == "GB43064" ~ "alpha5",
    ens_gene == "GB43416" ~ "alpha6",
    ens_gene == "GB53053" ~ "alpha7",
    ens_gene == "GB40923" ~ "alpha8",
    ens_gene == "GB53427" ~ "alpha9",
    ens_gene == "GB53055" ~ "beta1",
    ens_gene == "GB53428" ~ "beta2"
  )) %>%
  dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(contrast = "queen_vs_worker") %>%
  dplyr::mutate(species = "amel") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(castes_resruv_df,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "amel_castes_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)
```

# Analysis of tissues
## Create a `DESeqDataSet` object
```{r}
tissues_dds <- DESeq2::DESeqDataSetFromTximport(tissues_txi,
  colData = tissues_table,
  design = ~tissue
)

# Choose the reference level for tissue factor
tissues_dds$tissue <- relevel(tissues_dds$tissue, ref = "midgut")
```

## Pre-filter the low-count genes
```{r}
# Estimate the size factors
tissues_dds <- DESeq2::estimateSizeFactors(tissues_dds)

# Pre-filter the dataset
dim(tissues_dds)
# 15314    48
keep <- rowSums(DESeq2::counts(tissues_dds) >= 10) >= 3
tissues_dds <- tissues_dds[keep, ]
dim(tissues_dds)
# 13114    48
```

## Differential gene expression analysis
```{r}
# Run DESeq2 pipeline
tissues_dds <- DESeq2::DESeq(tissues_dds, test = "Wald", betaPrior = FALSE)
# DESeq2::resultsNames(tissues_dds)

tissues <- tissues_table %>%
  dplyr::pull(tissue) %>%
  unique() %>%
  as.character()

# Create list of condition pairs
tissues_contrasts <- combn(x = tissues, m = 2) %>%
  t() %>%
  as.data.frame.array() %>%
  dplyr::rename(cond1 = V1, cond2 = V2) %>%
  dplyr::mutate(contrast = paste0(cond1, "_vs_", cond2))
tissues_contrasts

# Extract results
tissues_res <- list()
for (row in 1:nrow(tissues_contrasts)) {
  cond1 <- as.character(tissues_contrasts[row, "cond1"])
  cond2 <- as.character(tissues_contrasts[row, "cond2"])
  tissues_res[[row]] <- DESeq2::results(
    tissues_dds,
    contrast = c("tissue", cond1, cond2),
    lfcThreshold = 0,
    alpha = 0.1,
    pAdjustMethod = "BH"
  )
}

names(tissues_res) <- tissues_contrasts$contrast
```

## Create table of `DESeq2` Wald test results for tissues
```{r}
tissues_res_df <- list()
for (contrast in tissues_contrasts$contrast) {
  tissues_res_df[[contrast]] <- tissues_res[[contrast]] %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "ens_gene") %>%
    dplyr::filter(ens_gene %in% amel_nachrs) %>%
    dplyr::mutate(subunit = case_when(
      ens_gene == "GB42850" ~ "alpha1",
      ens_gene == "GB42644" ~ "alpha2",
      ens_gene == "GB47845" ~ "alpha3",
      ens_gene == "GB43275" ~ "alpha4",
      ens_gene == "GB43064" ~ "alpha5",
      ens_gene == "GB43416" ~ "alpha6",
      ens_gene == "GB53053" ~ "alpha7",
      ens_gene == "GB40923" ~ "alpha8",
      ens_gene == "GB53427" ~ "alpha9",
      ens_gene == "GB53055" ~ "beta1",
      ens_gene == "GB53428" ~ "beta2"
    )) %>%
    dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
    dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
    dplyr::mutate(contrast = contrast) %>%
    dplyr::mutate(species = "amel") %>%
    dplyr::mutate_if(is.numeric, round, digits = 4)
}

tissues_res_df <- dplyr::bind_rows(tissues_res_df)

# Write table of `DESeq2` Wald test results
write.csv(tissues_res_df,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "amel_tissues_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)

# table(tissues_res_df$significance)
# table(tissues_res_df$subunit, tissues_res_df$significance)
```

## Perform likelihood ratio test for testing all tissues at once
```{r}
tissues_lrt <- DESeq2::DESeq(tissues_dds, test = "LRT", reduced = ~1)
tissues_lrt_res <- DESeq2::results(tissues_lrt)
tissues_lrt_res_df <- tissues_lrt_res %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% amel_nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "GB42850" ~ "alpha1",
    ens_gene == "GB42644" ~ "alpha2",
    ens_gene == "GB47845" ~ "alpha3",
    ens_gene == "GB43275" ~ "alpha4",
    ens_gene == "GB43064" ~ "alpha5",
    ens_gene == "GB43416" ~ "alpha6",
    ens_gene == "GB53053" ~ "alpha7",
    ens_gene == "GB40923" ~ "alpha8",
    ens_gene == "GB53427" ~ "alpha9",
    ens_gene == "GB53055" ~ "beta1",
    ens_gene == "GB53428" ~ "beta2"
  )) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(species = "amel") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(tissues_lrt_res_df,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "amel_tissues_deseq2_lrt_res_nachrs.csv"
  ),
  row.names = FALSE
)
```

# Analysis of foragers and nurses
```{r}
# Combine the factors of interest (tissue and behavior) into a single factor
tissues_dds$group <- factor(paste0(
  tissues_dds$behavior, tissues_dds$tissue
))

# Change the design to include only this factor, e.g. ~ group
design(tissues_dds) <- ~group

# Run the DESeq2 pipeline
tissues_dds <- DESeq2::DESeq(tissues_dds, test = "Wald", betaPrior = FALSE)

# Create list of group pairs, comparing the same tissue from
# foragers and nurses
tissues_foragers <- tissues_table %>%
  dplyr::filter(behavior == "forager") %>%
  dplyr::mutate(group = paste0(behavior, tissue)) %>%
  dplyr::pull(group) %>%
  unique() %>% 
  sort()

tissues_nurses <- tissues_table %>%
  dplyr::filter(behavior == "nurse") %>%
  dplyr::mutate(group = paste0(behavior, tissue)) %>%
  dplyr::pull(group) %>%
  unique() %>% 
  sort()

groups_contrasts <- dplyr::tibble(
  group1 = tissues_foragers,
  group2 = tissues_nurses,
  contrast = paste0(group1, "_vs_", group2)
  )

# Extract results
groups_res <- list()
for (row in 1:nrow(groups_contrasts)) {
  group1 <- as.character(groups_contrasts[row, "group1"])
  group2 <- as.character(groups_contrasts[row, "group2"])
  groups_res[[row]] <- DESeq2::results(
    tissues_dds,
    contrast = c("group", group1, group2),
    lfcThreshold = 0,
    alpha = 0.1,
    pAdjustMethod = "BH"
  )
}

names(groups_res) <- groups_contrasts$contrast
```

## Create table of `DESeq2` Wald test results for foragers and nurses
```{r}
groups_res_df <- list()
for (contrast in groups_contrasts$contrast) {
  groups_res_df[[contrast]] <- groups_res[[contrast]] %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "ens_gene") %>%
    dplyr::filter(ens_gene %in% amel_nachrs) %>%
    dplyr::mutate(subunit = case_when(
      ens_gene == "GB42850" ~ "alpha1",
      ens_gene == "GB42644" ~ "alpha2",
      ens_gene == "GB47845" ~ "alpha3",
      ens_gene == "GB43275" ~ "alpha4",
      ens_gene == "GB43064" ~ "alpha5",
      ens_gene == "GB43416" ~ "alpha6",
      ens_gene == "GB53053" ~ "alpha7",
      ens_gene == "GB40923" ~ "alpha8",
      ens_gene == "GB53427" ~ "alpha9",
      ens_gene == "GB53055" ~ "beta1",
      ens_gene == "GB53428" ~ "beta2"
    )) %>%
    dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
    dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
    dplyr::mutate(contrast = contrast) %>%
    dplyr::mutate(species = "amel") %>%
    dplyr::mutate_if(is.numeric, round, digits = 4)
}

groups_res_df <- dplyr::bind_rows(groups_res_df)

# Write table of `DESeq2` Wald test results
write.csv(groups_res_df,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "amel_foragers_nurses_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)
```

# Supplementary materials: data visualisation
## Principal component analysis of samples from tissues
```{r}
# Apply the variance stabilizing transformation
# The transformed data are on the log2 scale
tissues_vsd <- DESeq2::vst(tissues_dds, blind = FALSE)

tissues_pca_data <- DESeq2::plotPCA(tissues_vsd,
  intgroup = "tissue",
  ntop = 500, returnData = TRUE
)

percent_var <- round(100 * attr(tissues_pca_data, "percentVar"))

ggplot(tissues_pca_data, aes(
  x = PC1, y = PC2,
  color = tissue, shape = tissue
)) +
  geom_point(
    size = 4, stroke = 1
  ) +
  xlab(paste0("PC1 [", percent_var[1], "%]")) +
  ylab(paste0("PC2 [", percent_var[2], "%]")) +
  coord_fixed(ratio = 1) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "plain"),
    plot.subtitle = element_text(hjust = 0.5, size = 14, face = "plain"),
    legend.title = element_blank(),
    legend.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text = element_text(size = 12)
  ) +
  scale_shape_manual(values = 1:9) +
  labs(
    title = "Relationships among samples from honey bee tissues"
  )

ggsave(
  filename = here::here(
    "results", "2020-07-28-gene_expression", "supplementary",
    "amel_tissues_vst_counts_pca.pdf"
  ),
  width = 8,
  height = 6,
  units = "in"
)
```

## Heatmap of transformed, normalised counts of samples from tissues
```{r}
tissues_vsd_nachrs <- assay(tissues_vsd) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  tidyr::gather(key = "sample", value = "vst_count", -ens_gene) %>%
  dplyr::mutate_if(is.numeric, round, digits = 4) %>%
  dplyr::filter(ens_gene %in% amel_nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "GB42850" ~ "alpha1",
    ens_gene == "GB42644" ~ "alpha2",
    ens_gene == "GB47845" ~ "alpha3",
    ens_gene == "GB43275" ~ "alpha4",
    ens_gene == "GB43064" ~ "alpha5",
    ens_gene == "GB43416" ~ "alpha6",
    ens_gene == "GB53053" ~ "alpha7",
    ens_gene == "GB40923" ~ "alpha8",
    ens_gene == "GB53427" ~ "alpha9",
    ens_gene == "GB53055" ~ "beta1",
    ens_gene == "GB53428" ~ "beta2"
  ))

regex_remove <- c("^.+?_", "_\\w\\d$")
tissues_vsd_nachrs <- tissues_vsd_nachrs%>%
  dplyr::mutate(group = stringr::str_remove_all(sample,
    paste(regex_remove, collapse = "|"))) %>%
  dplyr::mutate(
    group = stringr::str_to_sentence(group)) %>%
  dplyr::mutate(
    group = stringr::str_replace(group, "_", " "))

ggplot(tissues_vsd_nachrs, aes(
  x = sample, y = subunit, fill = vst_count
)) +
  geom_tile(color = "white") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "plain"),
    plot.subtitle = element_text(hjust = 0.5, size = 14, face = "plain"),
    legend.title = element_blank(),
    legend.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text = element_text(size = 10),
    # axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
    axis.text.x = element_blank(),
    panel.spacing = unit(4, "mm"),
    strip.text = element_text(size = 10)
  ) +
  labs(
    title = "Heatmap of transformed counts from honey bee tissues",
    x = "Sample",
    y = expression(paste("Log"[2], " (VST) transformed count"))
  ) +
  scale_y_discrete(limits = rev(levels(factor(tissues_vsd_nachrs$subunit)))) +
  scale_fill_gradientn(colours = c("#3B9AB2", "#78B7C5", "#EBCC2A")) +
  facet_wrap(~group, ncol = 10, scales = "free_x", 
    labeller = label_wrap_gen(12))

ggsave(
  filename = here::here(
    "results", "2020-07-28-gene_expression", "supplementary",
    "amel_tissues_vst_counts_heatmap.pdf"
  ),
  width = 12,
  height = 4,
  units = "in"
)
```
