---
title: "Differential expression analysis of nAChR subunits in _Bombus terrestris_"
author: "Federico Lopez"
date: '`r Sys.Date()`'
output:
  github_document:
    toc: yes
  pdf_document: 
    fig_caption: yes
    toc: yes
  html_document:
    toc: yes
editor_options: 
  chunk_output_type: console
geometry: margin = 1cm
---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  include = FALSE,
  out.height = "\textheight",
  out.width = "\textwidth"
)
```

# Load libraries
```{r}
load_pkgs <- function(pkg) {
  sapply(pkg, require, character.only = TRUE)
}

cran_pkgs <- c(
  "BiocManager", "tidyverse", "RColorBrewer", "pheatmap", "styler", "here"
)
load_pkgs(cran_pkgs)

bioconductor_pkgs <- c(
  "biomaRt", "rhdf5", "tximport", "DESeq2"
)
load_pkgs(bioconductor_pkgs)
```

# Set a custom ggplot theme
```{r}
# Set a custom ggplot theme
# devtools::install_github("cttobin/ggthemr")
library(ggthemr)
custom_palette <- c(
  "#9E9E9E", "#59BDEF", "#EBCC2A", "#E1AF00", "#F27C8D", "#7B73D1", "#7B9FE0",
  "#F5CC7F", "#66C2A5", "#28A7B6", "#F2CB57", "#F2A057", "#F2845C"
)
# ggthemr::colour_plot(custom_palette)

custom_theme <- ggthemr::define_palette(
  swatch = custom_palette,
  gradient = c(lower = "#59BDEF", upper = "#FF6B5A")
)
ggthemr::ggthemr(custom_theme)
```

# Create directory for results
```{r}
# here::here()
# Create directory for results
dir.create(here::here("results", "2020-07-28-gene_expression"),
  recursive = TRUE
)
```

# Associate transcripts to genes
```{r}
# Remove all biomaRt cached files
# biomaRt::biomartCacheInfo()
# biomaRt::biomartCacheClear()

# Retrieve Ensembl gene identifiers/names and create transcript-to-gene table
biomaRt::listMarts(host = "metazoa.ensembl.org")

bter_mart <- biomaRt::useMart(
  biomart = "metazoa_mart",
  dataset = "bterrestris_eg_gene",
  host = "metazoa.ensembl.org" # version 48
)

bter_tx2gene <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
  mart = bter_mart
)
```

# Create vector of nAChR gene identifiers
```{r}
bter_nachrs <- c(
  "LOC100643274", "LOC100643282", "LOC100645032", "LOC100646787",
  "LOC100647301", "LOC100647350", "LOC100647624", "LOC100648987",
  "LOC100649515", "LOC100649612", "LOC100649796"
)
```

# Analysis of castes
## Read table of samples and covariates and `kallisto` results
```{r}
# Set sample identifiers using the names of the kallisto output directories
castes_ids <- dir(here::here(
  "2019-colgan_queen_worker_head", "2020-07-14-kallisto", "results"
))

# Read metadata table
castes_table <- read.csv(
  here::here(
    "bter_metadata",
    "2019-colgan_queen_worker_head_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

castes_table$caste <- relevel(
  castes_table$caste,
  ref = "worker"
)

castes_files <- here::here(
  "2019-colgan_queen_worker_head", "2020-07-14-kallisto", "results",
  castes_table$sample,
  "abundance.h5"
)

names(castes_files) <- castes_table$sample

# Check that sample names match
all(castes_ids %in% castes_table$sample)

# Check that all files exist
all(file.exists(castes_files))
```

## Create a `DESeqDataSet` object
```{r}
# Import kallisto quantifications
castes_txi <- tximport::tximport(castes_files,
  type = "kallisto",
  tx2gene = bter_tx2gene,
  txOut = FALSE
)

# Create a DESeqDataSet object
# The design below includes the batch factors "block" and "room"
castes_dds <- DESeq2::DESeqDataSetFromTximport(castes_txi,
  colData = castes_table,
  design = ~ block + room + caste
)

# Choose the reference level for caste factor
castes_dds$caste <- relevel(castes_dds$caste, ref = "worker")
```

## Pre-filter the low-count genes
```{r}
# Estimate the size factors
castes_dds <- DESeq2::estimateSizeFactors(castes_dds)

# Pre-filter the dataset
dim(castes_dds)
# 10661
# At least 4 samples with a count of 10 or higher
# The number of samples would be set to the smallest group size
keep <- rowSums(DESeq2::counts(castes_dds) >= 10) >= 4
castes_dds <- castes_dds[keep, ]
dim(castes_dds)
# 9139
```

## Differential gene expression analysis
```{r}
# Run DESeq2 pipeline
castes_dds <- DESeq2::DESeq(castes_dds, test = "Wald", betaPrior = FALSE)
DESeq2::resultsNames(castes_dds)

# Extract results
castes_res <- DESeq2::results(
  castes_dds,
  contrast = c("caste", "queen", "worker"),
  lfcThreshold = 0,
  alpha = 0.1,
  pAdjustMethod = "BH"
)

DESeq2::summary(castes_res)
```

## Create table of `DESeq2` Wald test results for castes
```{r}
castes_res <- castes_res %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% bter_nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "LOC100647624" ~ "alpha1",
    ens_gene == "LOC100647301" ~ "alpha2",
    ens_gene == "LOC100647350" ~ "alpha3",
    ens_gene == "LOC100645032" ~ "alpha4",
    ens_gene == "LOC100648987" ~ "alpha4e5",
    ens_gene == "LOC100649515" ~ "alpha5",
    ens_gene == "LOC100643274" ~ "alpha6",
    ens_gene == "LOC100649796" ~ "alpha7",
    ens_gene == "LOC100643282" ~ "alpha8",
    ens_gene == "LOC100649612" ~ "beta1",
    ens_gene == "LOC100646787" ~ "beta2"
  )) %>%
  dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(contrast = "queen_vs_worker") %>%
  dplyr::mutate(study = "colgan") %>%
  dplyr::mutate(species = "bter") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(castes_res,
  file = here::here(
    "results", "2020-07-28-gene_expression", 
    "bter_castes_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)

table(castes_res$significance)
# notsig    sig 
#      4      6

table(castes_res$subunit, castes_res$significance)
#          notsig sig
# alpha1        1   0
# alpha2        0   1
# alpha3        0   1
# alpha4e5      1   0
# alpha5        0   1
# alpha6        0   1
# alpha7        0   1
# alpha8        0   1
# beta1         1   0
# beta2         1   0
```

# Analysis of life stages
## Read `kallisto` results and table of samples and covariates.
```{r}
# Set sample identifiers using the names of the kallisto output directories
lifestages_ids <- dir(here::here(
  "2015-harrison_queen_worker_whole_body", "results",
  "2020-07-14-kallisto"
))

# Read metadata table
lifestages_table <- read.csv(
  here::here(
    "bter_metadata",
    "2015-harrison_queen_worker_whole_body_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

lifestages_table$condition <- relevel(
  lifestages_table$condition,
  ref = "worker_larva"
)

lifestages_files <- here::here(
  "2015-harrison_queen_worker_whole_body",
  "2020-07-14-kallisto", "results",
  lifestages_table$sample,
  "abundance.h5"
)

names(lifestages_files) <- lifestages_table$sample

# Check that sample names match
all(lifestages_ids %in% lifestages_table$sample)

# Check that all files exist
all(file.exists(lifestages_files))
```

## Create a `DESeqDataSet` object
```{r}
# Import kallisto quantifications
lifestages_txi <- tximport::tximport(lifestages_files,
  type = "kallisto",
  tx2gene = bter_tx2gene,
  txOut = FALSE
)

# Create a DESeqDataSet object
lifestages_dds <- DESeq2::DESeqDataSetFromTximport(lifestages_txi,
  colData = lifestages_table,
  design = ~condition
)

# Choose the reference level for condition factor
lifestages_dds$condition <- relevel(lifestages_dds$condition, 
  ref = "worker_larva")
```

## Pre-filter the low-count genes
```{r}
# Estimate the size factors
lifestages_dds <- DESeq2::estimateSizeFactors(lifestages_dds)

# Pre-filter the dataset
dim(lifestages_dds)
# 10661
# At least 3 samples with a count of 10 or higher
# The number of samples would be set to the smallest group size
keep <- rowSums(DESeq2::counts(lifestages_dds) >= 10) >= 3
lifestages_dds <- lifestages_dds[keep, ]
dim(lifestages_dds)
# 9892
```

## Differential gene expression analysis
```{r}
# Run DESeq2 pipeline
lifestages_dds <- DESeq2::DESeq(lifestages_dds, test = "Wald", 
  betaPrior = FALSE)
DESeq2::resultsNames(lifestages_dds)

lifestages <- lifestages_table %>%
  dplyr::pull(condition) %>%
  unique() %>%
  as.character()

# Create list of condition pairs
lifestages_contrasts <- combn(x = lifestages, m = 2) %>%
  t() %>%
  as.data.frame.array() %>%
  dplyr::rename(cond1 = V1, cond2 = V2) %>%
  dplyr::mutate(contrast = paste0(cond1, "_vs_", cond2))

# Extract results
lifestages_res <- list()
for (row in 1:nrow(lifestages_contrasts)) {
  cond1 <- as.character(lifestages_contrasts[row, "cond1"])
  cond2 <- as.character(lifestages_contrasts[row, "cond2"])
  lifestages_res[[row]] <- DESeq2::results(
    lifestages_dds,
    contrast = c("condition", cond1, cond2),
    lfcThreshold = 0,
    alpha = 0.1,
    pAdjustMethod = "BH"
  )
}

names(lifestages_res) <- lifestages_contrasts$contrast
```

## Create table of `DESeq2` Wald test results for life stages
```{r}
lifestages_res_df <- list()
for (contrast in lifestages_contrasts$contrast) {
  lifestages_res_df[[contrast]] <- lifestages_res[[contrast]] %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "ens_gene") %>%
    dplyr::filter(ens_gene %in% bter_nachrs) %>%
    dplyr::mutate(subunit = case_when(
      ens_gene == "LOC100647624" ~ "alpha1",
      ens_gene == "LOC100647301" ~ "alpha2",
      ens_gene == "LOC100647350" ~ "alpha3",
      ens_gene == "LOC100645032" ~ "alpha4",
      ens_gene == "LOC100648987" ~ "alpha4e5",
      ens_gene == "LOC100649515" ~ "alpha5",
      ens_gene == "LOC100643274" ~ "alpha6",
      ens_gene == "LOC100649796" ~ "alpha7",
      ens_gene == "LOC100643282" ~ "alpha8",
      ens_gene == "LOC100649612" ~ "beta1",
      ens_gene == "LOC100646787" ~ "beta2"
    )) %>%
    dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
    dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
    dplyr::mutate(contrast = contrast) %>%
    dplyr::mutate(study = "harrison") %>%
    dplyr::mutate(species = "bter") %>%
    dplyr::mutate_if(is.numeric, round, digits = 4)
}

lifestages_res_df <- dplyr::bind_rows(lifestages_res_df)

# Write table of `DESeq2` Wald test results
write.csv(lifestages_res_df,
  file = here::here(
    "results", "2020-07-28-gene_expression", 
    "bter_life_stages_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)

table(lifestages_res_df$significance)
# not_sig     sig
#     40      20

table(lifestages_res_df$subunit, lifestages_res_df$significance)
#          notsig sig
# alpha1        4   2
# alpha2        3   3
# alpha3        4   2
# alpha4e5      3   3
# alpha5        4   2
# alpha6        5   1
# alpha7        5   1
# alpha8        6   0
# beta1         4   2
# beta2         2   4
```

## Perform likelihood ratio test for testing all life stages at once
```{r}
lifestages_lrt <- DESeq2::DESeq(lifestages_dds, test = "LRT", reduced = ~1)
lifestages_lrt_res <- DESeq2::results(lifestages_lrt)
lifestages_lrt_res_df <- lifestages_lrt_res %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% bter_nachrs) %>%
  dplyr::mutate(subunit = case_when(
      ens_gene == "LOC100647624" ~ "alpha1",
      ens_gene == "LOC100647301" ~ "alpha2",
      ens_gene == "LOC100647350" ~ "alpha3",
      ens_gene == "LOC100645032" ~ "alpha4",
      ens_gene == "LOC100648987" ~ "alpha4e5",
      ens_gene == "LOC100649515" ~ "alpha5",
      ens_gene == "LOC100643274" ~ "alpha6",
      ens_gene == "LOC100649796" ~ "alpha7",
      ens_gene == "LOC100643282" ~ "alpha8",
      ens_gene == "LOC100649612" ~ "beta1",
      ens_gene == "LOC100646787" ~ "beta2"
  )) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(species = "bter") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(lifestages_lrt_res_df,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "bter_life_stages_deseq2_lrt_res_nachrs.csv"
  ),
  row.names = FALSE
)
```

# Combine results into a single data frame
```{r}
bter_res_nachrs <- dplyr::bind_rows(castes_res, lifestages_res_df)

write.csv(bter_res_nachrs,
  file = here::here(
    "results", "2020-07-28-gene_expression",
    "bter_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)
```
