---
title: "Differential expression analysis of nAChR subunits in _Bombus terrestris_"
author: "Federico Lopez"
date: '`r Sys.Date()`'
output:
  github_document:
    toc: yes
  pdf_document: 
    fig_caption: yes
    toc: yes
  html_document:
    toc: yes
editor_options: 
  chunk_output_type: console
geometry: margin = 1cm
---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  include = FALSE,
  out.height = "\textheight",
  out.width = "\textwidth"
)
```

# Load libraries
```{r}
load_pkgs <- function(pkg) {
  sapply(pkg, require, character.only = TRUE)
}

cran_pkgs <- c(
  "BiocManager", "tidyverse", "GGally", "ggtext", "ggh4x", "ggforce",
  "RColorBrewer", "styler", "patchwork", "here"
)

load_pkgs(cran_pkgs)

bioconductor_pkgs <- c(
  "biomaRt", "rhdf5", "tximport", "DESeq2", "vsn", "hexbin"
)

load_pkgs(bioconductor_pkgs)
```

# Set a custom `ggplot` theme
```{r}
# Set a custom ggplot theme
# devtools::install_github("cttobin/ggthemr")
library(ggthemr)

# Diverging
custom_palette <- c(
  "#9E9E9E", "#88CCEE", "#18913D", "#E9C83A", "#882255", "#CC6677",
  "#332288", "#7B73D1", "#2857BA", "#28A7B6", "#F2CB57", "#F2845C"
)

# ggthemr::colour_plot(custom_palette)

custom_theme <- ggthemr::define_palette(
  swatch = custom_palette,
  gradient = c(lower = "#E9C83A", upper = "#882255")
)

ggthemr::ggthemr(custom_theme)
```

```{r}
# here::here()
# Create directory for results
dir.create(here::here("results", "2022-07-20-gene_expression", "supplementary"),
  recursive = TRUE
)
```

# Associate transcripts to genes
```{r}
mart_bter <- biomaRt::useMart(
  biomart = "metazoa_mart",
  dataset = "bterrestris_eg_gene",
  host = "sep2019-metazoa.ensembl.org"
)

tx2gene_bter <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
  mart = mart_bter
)

# tx2gene_bter <- read.csv(file = "bter1.0tx2gene.csv", 
#   header = TRUE)
```

# Create vector of nAChR gene identifiers
```{r}
nachrs_bter <- c(
  "LOC100643274", "LOC100643282", "LOC100645032", "LOC100646787",
  "LOC100647301", "LOC100647350", "LOC100647624", "LOC100648987",
  "LOC100649515", "LOC100649612", "LOC100649796"
)
```

# Analysis of castes
## Read table of samples and covariates and `kallisto` results
```{r}
# Set sample identifiers using the names of the kallisto output directories
ids_castes_bter <- dir(here::here(
  "results", "2019-colgan_queen_worker_head", "2020-07-14-kallisto", "results"
))

# Read metadata table
table_castes_bter <- read.csv(
  here::here(
    "metadata_bter",
    "2019-colgan_queen_worker_head_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

table_castes_bter$caste <- relevel(
  table_castes_bter$caste,
  ref = "worker"
)

files_castes_bter <- here::here(
  "results", "2019-colgan_queen_worker_head", "2020-07-14-kallisto", "results",
  table_castes_bter$sample,
  "abundance.h5"
)

names(files_castes_bter) <- table_castes_bter$sample

# Check that sample names match
all(ids_castes_bter %in% table_castes_bter$sample)

# Check that all files exist
all(file.exists(files_castes_bter))
```

## Create a `DESeqDataSet` object
```{r}
# Import kallisto quantifications
txi_castes_bter <- tximport::tximport(files_castes_bter,
  type = "kallisto",
  tx2gene = tx2gene_bter,
  txOut = FALSE
)

# Create a DESeqDataSet object
# The design below includes the batch factors "block" and "room"
dds_castes_bter <- DESeq2::DESeqDataSetFromTximport(txi_castes_bter,
  colData = table_castes_bter,
  design = ~ block + room + caste
)

# Choose the reference level for caste factor
dds_castes_bter$caste <- relevel(dds_castes_bter$caste, ref = "worker")

abundance_castes_bter <- txi_castes_bter$abundance %>% 
  as.data.frame() %>%
  tibble::rownames_to_column(var = "EnsemblGeneID")

write.csv(abundance_castes_bter,
  file = here::here(
    "results", "2022-07-20-gene_expression", 
    "ensembl_bter1.0_tximport_abundance_castes_bter.csv"),
  quote = FALSE, row.names = FALSE)
```

## Pre-filter the low-count genes
```{r}
# Estimate the size factors
dds_castes_bter <- DESeq2::estimateSizeFactors(dds_castes_bter)

# Pre-filter the dataset
dim(dds_castes_bter)
# 10661
# At least 4 samples with a count of 10 or higher
# The number of samples would be set to the smallest group size
keep <- rowSums(DESeq2::counts(dds_castes_bter) >= 10) >= 4
dds_castes_bter <- dds_castes_bter[keep, ]
dim(dds_castes_bter)
# 9139
```

## Differential gene expression analysis
```{r}
# Run DESeq2 pipeline
dds_castes_bter <- DESeq2::DESeq(dds_castes_bter, test = "Wald", betaPrior = FALSE)

DESeq2::resultsNames(dds_castes_bter)

# Extract results
res_castes_bter <- DESeq2::results(
  dds_castes_bter,
  contrast = c("caste", "queen", "worker"),
  lfcThreshold = 0,
  alpha = 0.1,
  pAdjustMethod = "BH"
)

DESeq2::summary(res_castes_bter)
```

## Create table of `DESeq2` Wald test results for castes
```{r}
res_castes_bter <- res_castes_bter %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% nachrs_bter) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "LOC100647624" ~ "alpha1",
    ens_gene == "LOC100647301" ~ "alpha2",
    ens_gene == "LOC100647350" ~ "alpha3",
    ens_gene == "LOC100645032" ~ "alpha4",
    ens_gene == "LOC100648987" ~ "alpha4e5",
    ens_gene == "LOC100649515" ~ "alpha5",
    ens_gene == "LOC100643274" ~ "alpha6",
    ens_gene == "LOC100649796" ~ "alpha7",
    ens_gene == "LOC100643282" ~ "alpha8",
    # ens_gene == "LOC100646665" ~ "alpha9", # Missing on Ensembl45
    ens_gene == "LOC100649612" ~ "beta1",
    ens_gene == "LOC100646787" ~ "beta2"
  )) %>%
  dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(contrast = "queen_vs_worker") %>%
  dplyr::mutate(study = "colgan") %>%
  dplyr::mutate(species = "bter") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(res_castes_bter,
  file = here::here(
    "results", "2022-07-20-gene_expression", 
    "bter_castes_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)

table(res_castes_bter$significance)
# notsig    sig 
#      4      6

table(res_castes_bter$subunit, res_castes_bter$significance)
#          notsig sig
# alpha1        1   0
# alpha2        0   1
# alpha3        0   1
# alpha4e5      1   0
# alpha5        0   1
# alpha6        0   1
# alpha7        0   1
# alpha8        0   1
# beta1         1   0
# beta2         1   0
```

# Analysis of life stages
## Read `kallisto` results and table of samples and covariates.
```{r}
# Set sample identifiers using the names of the kallisto output directories
ids_lifestages_bter <- dir(here::here(
  "results", "2015-harrison_queen_worker_whole_body", "results",
  "2020-07-14-kallisto"
))

# Read metadata table
table_lifestages_bter <- read.csv(
  here::here(
    "metadata_bter",
    "2015-harrison_queen_worker_whole_body_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

table_lifestages_bter$condition <- relevel(
  table_lifestages_bter$condition,
  ref = "worker_larva"
)

files_lifestages_bter <- here::here(
  "results", "2015-harrison_queen_worker_whole_body",
  "2020-07-14-kallisto", "results",
  table_lifestages_bter$sample,
  "abundance.h5"
)

names(files_lifestages_bter) <- table_lifestages_bter$sample

# Check that sample names match
all(ids_lifestages_bter %in% table_lifestages_bter$sample)

# Check that all files exist
all(file.exists(files_lifestages_bter))
```

## Create a `DESeqDataSet` object
```{r}
# Import kallisto quantifications
txi_lifestages_bter <- tximport::tximport(files_lifestages_bter,
  type = "kallisto",
  tx2gene = tx2gene_bter,
  txOut = FALSE
)

# Create a DESeqDataSet object
dds_lifestages_bter <- DESeq2::DESeqDataSetFromTximport(txi_lifestages_bter,
  colData = table_lifestages_bter,
  design = ~condition
)

# Choose the reference level for condition factor
dds_lifestages_bter$condition <- relevel(dds_lifestages_bter$condition, 
  ref = "worker_larva")

abundance_lifestages_bter <- txi_lifestages_bter$abundance %>% 
  as.data.frame() %>%
  tibble::rownames_to_column(var = "EnsemblGeneID")

write.csv(abundance_lifestages_bter,
  file = here::here(
    "results", "2022-07-20-gene_expression", 
    "ensembl_bter1.0_tximport_abundance_lifestages_bter.csv"),
  quote = FALSE, row.names = FALSE)
```

## Pre-filter the low-count genes
```{r}
# Estimate the size factors
dds_lifestages_bter <- DESeq2::estimateSizeFactors(dds_lifestages_bter)

# Pre-filter the dataset
dim(dds_lifestages_bter)
# 10661
# At least 3 samples with a count of 10 or higher
# The number of samples would be set to the smallest group size
keep <- rowSums(DESeq2::counts(dds_lifestages_bter) >= 10) >= 3
dds_lifestages_bter <- dds_lifestages_bter[keep, ]
dim(dds_lifestages_bter)
# 9892
```

## Differential gene expression analysis
```{r}
# Run DESeq2 pipeline
dds_lifestages_bter <- DESeq2::DESeq(dds_lifestages_bter, test = "Wald", 
  betaPrior = FALSE)

DESeq2::resultsNames(dds_lifestages_bter)

lifestages_bter <- table_lifestages_bter %>%
  dplyr::pull(condition) %>%
  unique() %>%
  as.character()

# Create list of condition pairs
contrasts_lifestages <- combn(x = lifestages_bter, m = 2) %>%
  t() %>%
  as.data.frame.array() %>%
  dplyr::rename(cond1 = V1, cond2 = V2) %>%
  dplyr::mutate(contrast = paste0(cond1, "_vs_", cond2))

contrasts_lifestages

# Extract results
res_lifestages <- list()

for (row in 1:nrow(contrasts_lifestages)) {
  cond1 <- as.character(contrasts_lifestages[row, "cond1"])
  cond2 <- as.character(contrasts_lifestages[row, "cond2"])
  res_lifestages[[row]] <- DESeq2::results(
    dds_lifestages_bter,
    contrast = c("condition", cond1, cond2),
    lfcThreshold = 0,
    alpha = 0.1,
    pAdjustMethod = "BH"
  )
}

names(res_lifestages) <- contrasts_lifestages$contrast
```

## Create table of `DESeq2` Wald test results for life stages
```{r}
res_lifestages_df <- list()

for (contrast in contrasts_lifestages$contrast) {
  res_lifestages_df[[contrast]] <- res_lifestages[[contrast]] %>%
    as.data.frame() %>%
    tibble::rownames_to_column(var = "ens_gene") %>%
    dplyr::filter(ens_gene %in% nachrs_bter) %>%
    dplyr::mutate(subunit = case_when(
      ens_gene == "LOC100647624" ~ "alpha1",
      ens_gene == "LOC100647301" ~ "alpha2",
      ens_gene == "LOC100647350" ~ "alpha3",
      ens_gene == "LOC100645032" ~ "alpha4",
      ens_gene == "LOC100648987" ~ "alpha4e5",
      ens_gene == "LOC100649515" ~ "alpha5",
      ens_gene == "LOC100643274" ~ "alpha6",
      ens_gene == "LOC100649796" ~ "alpha7",
      ens_gene == "LOC100643282" ~ "alpha8",
      ens_gene == "LOC100649612" ~ "beta1",
      ens_gene == "LOC100646787" ~ "beta2"
    )) %>%
    dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "up", "down")) %>%
    dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
    dplyr::mutate(contrast = contrast) %>%
    dplyr::mutate(study = "harrison") %>%
    dplyr::mutate(species = "bter") %>%
    dplyr::mutate_if(is.numeric, round, digits = 4)
}

res_lifestages_df <- dplyr::bind_rows(res_lifestages_df)

# Write table of `DESeq2` Wald test results
write.csv(res_lifestages_df,
  file = here::here(
    "results", "2022-07-20-gene_expression", 
    "bter_life_stages_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)

table(res_lifestages_df$significance)
# not_sig     sig
#     40      20

table(res_lifestages_df$subunit, res_lifestages_df$significance)
#          notsig sig
# alpha1        4   2
# alpha2        3   3
# alpha3        4   2
# alpha4e5      3   3
# alpha5        4   2
# alpha6        5   1
# alpha7        5   1
# alpha8        6   0
# beta1         4   2
# beta2         2   4
```

## Perform likelihood ratio test for testing all life stages at once
```{r}
lrt_lifestages <- DESeq2::DESeq(dds_lifestages_bter, test = "LRT", reduced = ~1)

res_lrt_lifestages <- DESeq2::results(lrt_lifestages)

res_lrt_lifestages_df <- res_lrt_lifestages %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% nachrs_bter) %>%
  dplyr::mutate(subunit = case_when(
      ens_gene == "LOC100647624" ~ "alpha1",
      ens_gene == "LOC100647301" ~ "alpha2",
      ens_gene == "LOC100647350" ~ "alpha3",
      ens_gene == "LOC100645032" ~ "alpha4",
      ens_gene == "LOC100648987" ~ "alpha4e5",
      ens_gene == "LOC100649515" ~ "alpha5",
      ens_gene == "LOC100643274" ~ "alpha6",
      ens_gene == "LOC100649796" ~ "alpha7",
      ens_gene == "LOC100643282" ~ "alpha8",
      ens_gene == "LOC100649612" ~ "beta1",
      ens_gene == "LOC100646787" ~ "beta2"
  )) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "sig", "notsig")) %>%
  dplyr::mutate(species = "bter") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(res_lrt_lifestages_df,
  file = here::here(
    "results", "2022-07-20-gene_expression",
    "bter_life_stages_deseq2_lrt_res_nachrs.csv"
  ),
  row.names = FALSE
)
```

# Combine results into a single data frame
```{r}
res_nachrs_bter <- dplyr::bind_rows(res_castes_bter, res_lifestages_df)

write.csv(res_nachrs_bter,
  file = here::here(
    "results", "2022-07-20-gene_expression",
    "bter_deseq2_wald_res_nachrs.csv"
  ),
  row.names = FALSE
)
```
