---
title: "Differential expression analysis of nAChR genes in _Bombus terrestris_ using `DESeq2`"
author: "Federico Lopez"
date: '`r Sys.Date()`'
output:
  github_document:
    toc: yes
  pdf_document: 
    fig_caption: yes
    toc: yes
  html_document:
    toc: yes
editor_options: 
  chunk_output_type: console
geometry: margin = 1cm
---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  include = FALSE,
  out.height = "\textheight",
  out.width = "\textwidth"
)
```

# Set up the project directory to use `renv`
```{r}
renv::init(
  project = "~/projects/2020_bombus_nachrs", bare = TRUE,
  restart = FALSE
)
renv::snapshot()
```

# Load libraries and input data
Load required libraries.
```{r}
# Check if packages are installed and install new packages
# if necessary, then load packages
load_cran_pkgs <- function(pkg) {
  new_pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new_pkg)) {
    install.packages(new_pkg, dependencies = TRUE)
  }
  sapply(pkg, require, character.only = TRUE)
}

# Load CRAN packages
cran_pkgs <- c(
  "BiocManager", "tidyverse", "ggrepel", "ggforce", "ggnewscale", "ggstance",
  "ggsignif", "gghighlight", "GGally", "ggfortify", "RColorBrewer",
  "wesanderson", "pheatmap", "renv", "styler"
)
load_cran_pkgs(cran_pkgs)

# Load Bioconductor packages
load_bioconductor_pkgs <- function(pkg) {
  new_pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  if (length(new_pkg)) {
    BiocManager::install(new_pkg, version = "3.10")
  }
  sapply(pkg, require, character.only = TRUE)
}

bioconductor_pkgs <- c(
  "biomaRt", "rhdf5", "tximport", "DESeq2", "apeglm", "sva", "limma", "vsn",
  "hexbin"
  # "ComplexHeatmap"
)
load_bioconductor_pkgs(bioconductor_pkgs)

# Set a custom ggplot theme
# devtools::install_github("cttobin/ggthemr")
library(ggthemr)
# Use colorblind-friendly palette
# http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/#a-colorblind-friendly-palette
cbp1 <- c(
  "#6A6A6A", "#E69F00", "#56B4E9", "#009E73",
  "#F0E442", "#0072B2", "#D55E00", "#CC79A7"
)

# ggthemr::colour_plot(cbp1)

custom_theme <- ggthemr::define_palette(
  swatch = cbp1,
  gradient = c(lower = "#6A6A6A", upper = "#56B4E9")
)
ggthemr::ggthemr(custom_theme)
# ggthemr::ggthemr_reset()
```

Set the base directory including `kallisto` results and read
table of samples and covariates.
```{r}
# Set directory that contains the active file as the working directory
# setwd(getSrcDirectory()[1]) # Use outside of RStudio
# Comment out before knitting in RStudio
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
base_dir <- getwd()

# Create directory for results
dir.create("results/colgan", recursive = TRUE)

# Set sample identifiers using the names of the kallisto output directories
sample_ids <- dir(file.path(
  base_dir,
  "2019_colgan_queen-worker_head/2020-05-06-kallisto"
))

# Read metadata table
samples_table <- read.csv(
  file.path(
    base_dir, "2019_colgan_queen-worker_head",
    "2019_colgan_queen-worker_head_metadata.csv"
  ),
  header = TRUE,
  stringsAsFactors = TRUE
)

colnames(samples_table)

samples_table$caste <- relevel(
  samples_table$caste,
  ref = "queen"
)

input_files <- file.path(
  base_dir,
  "2019_colgan_queen-worker_head/2020-05-06-kallisto",
  samples_table$sample,
  "abundance.h5"
)

names(input_files) <- samples_table$sample
input_files

# Check that all files exist
all(file.exists(input_files))
# input_files
```

# Associate transcripts to genes
```{r}
# Remove all biomaRt cached files
# biomaRt::biomartCacheInfo()
# biomaRt::biomartCacheClear()

# Retrieve Ensembl gene identifiers/names and create transcript-to-gene table
biomaRt::listMarts(host = "metazoa.ensembl.org")
# biomaRt::listMarts(host = "eg40-metazoa.ensembl.org") # version 40
# biomaRt::listEnsemblArchives()

metazoa_mart <- biomaRt::useMart(
  biomart = "metazoa_mart",
  dataset = "bterrestris_eg_gene",
  host = "metazoa.ensembl.org" # version 47
)

transcript_to_gene <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
  mart = metazoa_mart
)

# biomaRt::listAttributes(metazoa_mart)
```

# Create a `DESeqDataSet` object
```{r}
# Import kallisto quantifications
txi <- tximport::tximport(input_files,
  type = "kallisto",
  tx2gene = transcript_to_gene,
  txOut = FALSE
)

# Create a DESeqDataSet object
# The design below includes the batch factor "block"
dds <- DESeq2::DESeqDataSetFromTximport(txi,
  colData = samples_table,
  design = ~ block + ct.room + caste
)
```

# Pre-filtering of low-count genes
```{r}
# Show table of sample information
colData(dds)

# Show number of replicates per condition
with(colData(dds), table(caste))

# Estimate the size factors
dds <- DESeq2::estimateSizeFactors(dds)
# With tximport, we compute gene x sample normalization factors
DESeq2::normalizationFactors(dds) %>% head()

# Choose the reference level
dds$caste <- relevel(dds$caste, ref = "worker")

# Pre-filter the dataset
nrow(dds)
# 10661
# At least 4 samples with a count of 10 or higher
# The number of samples would be set to the smallest group size
keep <- rowSums(DESeq2::counts(dds) >= 10) >= 4
dds <- dds[keep, ]
nrow(dds)
# 9139
```

# Differential gene expression analysis
## Run `DESeq2` pipeline
```{r}
# Run DESeq2 pipeline
dds <- DESeq2::DESeq(dds, test = "Wald", betaPrior = FALSE)
DESeq2::resultsNames(dds)

# Extract results
res <- DESeq2::results(
  dds,
  contrast = c("caste", "queen", "worker"),
  lfcThreshold = 0,
  alpha = 0.1,
  pAdjustMethod = "BH"
)

DESeq2::summary(res)

lfc <- DESeq2::lfcShrink(
  dds,
  coef = "caste_queen_vs_worker",
  type = "apeglm",
  lfcThreshold = 0,
)

DESeq2::summary(lfc)

# Create data frames
lfc_df <- lfc %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

sig_df <- lfc %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(padj < 0.05) %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)
```

## Write tables of `DESeq2` Wald test results.
```{r}
# Write results tables
write.csv(lfc_df,
  file = "results/colgan/deseq2_wald_lfc.csv",
  row.names = FALSE
)

write.csv(sig_df,
  file = "results/colgan/deseq2_wald_lfc-sig.csv",
  row.names = FALSE
)

nachrs <- c(
  "LOC100643274", "LOC100643282", "LOC100645032", "LOC100646787",
  "LOC100647301", "LOC100647350", "LOC100647624", "LOC100648987",
  "LOC100649515", "LOC100649612", "LOC100649796"
)

res_nachrs <- res %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "LOC100643274" ~ "Alpha6",
    ens_gene == "LOC100643282" ~ "Alpha8",
    ens_gene == "LOC100645032" ~ "Alpha4",
    ens_gene == "LOC100646787" ~ "Beta2",
    ens_gene == "LOC100647301" ~ "Alpha2",
    ens_gene == "LOC100647350" ~ "Alpha3",
    ens_gene == "LOC100647624" ~ "Alpha1",
    ens_gene == "LOC100648987" ~ "Alpha4e5",
    ens_gene == "LOC100649515" ~ "Alpha5",
    ens_gene == "LOC100649612" ~ "Beta1",
    ens_gene == "LOC100649796" ~ "Alpha7",
  )) %>%
  dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "Up", "Down")) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "Sig", "NotSig")) %>%
  dplyr::mutate(contrast = "QueenWorker") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(res_nachrs,
  file = "results/colgan/deseq2_wald_res-nachrs.csv",
  row.names = FALSE
)

lfc_nachrs <- lfc %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  dplyr::filter(ens_gene %in% nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "LOC100643274" ~ "Alpha6",
    ens_gene == "LOC100643282" ~ "Alpha8",
    ens_gene == "LOC100645032" ~ "Alpha4",
    ens_gene == "LOC100646787" ~ "Beta2",
    ens_gene == "LOC100647301" ~ "Alpha2",
    ens_gene == "LOC100647350" ~ "Alpha3",
    ens_gene == "LOC100647624" ~ "Alpha1",
    ens_gene == "LOC100648987" ~ "Alpha4e5",
    ens_gene == "LOC100649515" ~ "Alpha5",
    ens_gene == "LOC100649612" ~ "Beta1",
    ens_gene == "LOC100649796" ~ "Alpha7",
  )) %>%
  dplyr::mutate(regulation = ifelse(log2FoldChange > 0, "Up", "Down")) %>%
  dplyr::mutate(significance = ifelse(padj < 0.05, "Sig", "NotSig")) %>%
  dplyr::mutate(contrast = "QueenWorker") %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

write.csv(lfc_nachrs,
  file = "results/colgan/deseq2_wald_lfc-nachrs.csv",
  row.names = FALSE
)
```

# Data transformation and quality assessment
## Transform count data using `DESeq2`'s `vst` function
The variance stabilizing transformation (`vst`) is used for visualization, 
clustering or other machine learning tasks.
```{r}
# Variance stabilizing transformation
vsd <- DESeq2::vst(dds, blind = FALSE)

# Effects of transformations on the variance
vsn::meanSdPlot(assay(vsd))

# In the variance stabilized data the standard deviation is roughly constant
# along the whole range of mean counts

# Note that "while a flat curve of the square root of variance over the mean
# may seem like the goal of such transformations, this may be unreasonable in
# the case of datasets with many true differences due to the experimental
# conditions."
ggsave(
  filename = "results/colgan/bterrestris_kallisto-deseq2_vst-meanSdPlot.pdf",
  width = 5,
  height = 4,
  units = "in"
)
```

## Remove batch effects using `limma`'s `removeBatchEffect()` function
```{r}
# Removes any shifts in the log2-scale expression data that can be
# explained by batch
# Here "design" relates to "treatment conditions to be preserved, usually the
# design matrix with all experimental factors other than the batch effects."
mod <- model.matrix(~ 0 + caste, data = samples_table)
mat <- assay(vsd)
mat <- limma::removeBatchEffect(mat, batch = vsd$block, design = mod)
assay(vsd) <- mat
```

## Plot the (VST) transformed counts with batch variation removed
```{r}
# Create a tidy data frame with the transformed counts
vst_nachrs <- assay(vsd) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ens_gene") %>%
  tidyr::gather(key = "sample", value = "vst_count", -ens_gene) %>%
  dplyr::filter(ens_gene %in% nachrs) %>%
  dplyr::mutate(subunit = case_when(
    ens_gene == "LOC100643274" ~ "Alpha6",
    ens_gene == "LOC100643282" ~ "Alpha8",
    ens_gene == "LOC100646665" ~ "Alpha9",
    ens_gene == "LOC100645032" ~ "Alpha4",
    ens_gene == "LOC100646787" ~ "Beta2",
    ens_gene == "LOC100647301" ~ "Alpha2",
    ens_gene == "LOC100647350" ~ "Alpha3",
    ens_gene == "LOC100647624" ~ "Alpha1",
    ens_gene == "LOC100648987" ~ "Alpha4e5", # Alpha-4_exon5
    ens_gene == "LOC100649515" ~ "Alpha5",
    ens_gene == "LOC100649612" ~ "Beta1",
    ens_gene == "LOC100649796" ~ "Alpha7",
  )) %>%
  dplyr::mutate(caste = case_when(
    grepl("worker", sample) ~ "Worker",
    grepl("queen", sample) ~ "Queen"
  )) %>%
  dplyr::mutate_if(is.numeric, round, digits = 4)

head(vst_nachrs)

# Boxplots of transformed counts
contrast_colors <- c("#E69F00", "#56B4E9")

ggplot(vst_nachrs, aes(x = caste, y = vst_count, color = caste, fill = caste)) +
  geom_boxplot(
    alpha = 0.6,
    width = 0.4,
    position = position_dodge(0.8),
    size = 1
  ) +
  geom_jitter(
    shape = 21,
    fill = "white",
    size = 2,
    stroke = 1,
    position = position_dodge(0.8)
  ) +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "plain"),
    plot.subtitle = element_text(hjust = 0.5, size = 14, face = "plain"),
    strip.text = element_text(size = 12, face = "plain", hjust = 0.5),
    axis.title = element_text(size = 12),
    # axis.text = element_text(size = 12),
    axis.text.x = element_blank(),
    # axis.text.x = element_text(size = 10, angle = 45, hjust = 1),
    panel.spacing.x = unit(2, "mm"),
    legend.position = "bottom"
  ) +
  labs(
    x = "Caste",
    y = expression(paste("Log"[2], " (VST) transformed count")) # batch variation removed
  ) +
  scale_fill_manual(name = "Caste", values = contrast_colors) +
  scale_color_manual(name = "Caste", values = contrast_colors) +
  facet_wrap(~subunit, ncol = 10) +
  # ylim(NA, 10.5) +
  # scale_y_continuous(breaks = seq(6, 20, 1)) +
  ggsignif::geom_signif(
    comparisons = list(c("Queen", "Worker")),
    map_signif_level = TRUE, color = "#6A6A6A"
    # size = 1, textsize = 4, vjust = 0, y_position = 10
  )

ggsave(
  filename = "results/colgan/bterrestris_kallisto_vst-nachrs_boxplots.pdf",
  width = 10,
  height = 5,
  units = "in"
)
```

## Heatmap of the count matrix and clustering
```{r}
# Estimate variance of normalized counts for each gene, select top 50
select <- order(rowVars(counts(dds, normalized = TRUE)),
  decreasing = TRUE
)[1:50]

annotation <- as.data.frame(colData(dds)[, c(
  "caste",
  "block"
)])

annotation
colnames(annotation) <- c("Caste", "Block")

# Set colors
caste_colors <- list("Caste" = c(queen = "#E69F00", worker = "#56B4E9"))
block_colors <- list("Block" = c(block_1 = "#009E73", block_2 = "#F0E442"))
annot_colors <- c(caste_colors, block_colors)

# Plot the top 50 genes using the VST counts
vsd_pheatmap <- pheatmap::pheatmap(assay(vsd)[select, ],
  border_color = NA,
  cluster_rows = FALSE,
  cluster_cols = TRUE,
  show_rownames = TRUE,
  show_colnames = FALSE,
  annotation_col = annotation,
  fontsize = 10,
  fontsize_row = 8,
  fontface = "plain",
  treeheight_col = 30,
  annotation_colors = annot_colors
)

ggsave(vsd_pheatmap,
  filename = "results/colgan/bterrestris_kallisto-deseq2_vst-counts_top50_pheatmap.pdf",
  width = 4,
  height = 6,
  units = "in"
)
```

## Principal component plot of the samples
```{r}
pca_data <- DESeq2::plotPCA(vsd,
  intgroup = "caste",
  n = 500,
  returnData = TRUE
)

# Customize PCA plot
percent_var <- round(100 * attr(pca_data, "percentVar"))

ggplot(pca_data, aes(x = PC1, y = PC2, color = caste)) +
  geom_point(size = 4) +
  xlab(paste0("PC1 [", percent_var[1], "%]")) +
  ylab(paste0("PC2 [", percent_var[2], "%]")) +
  coord_fixed(ratio = 1) +
  theme(
    legend.title = element_blank(),
    legend.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    axis.text = element_text(size = 12)
  )

ggsave(
  filename = "results/colgan/bterrestris_kallisto-deseq2_vst-pca.pdf",
  width = 6,
  height = 4,
  units = "in"
)
```
