---
title: "nAChRs expression in queens and workers of Apis mellifera"
author: "Alicja Witwicka"
date: "23.04.2020"
output:
  github_document:
    toc: yes
  pdf_document: 
    fig_caption: yes
    toc: yes
  html_document:
    toc: yes
number_sections: true
number_sections: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.path = "")
```
```{r libraries, echo = FALSE, message = FALSE, warning = FALSE}
#renv::snapshot()
```
```{r libraries, echo = FALSE, message = FALSE, warning = FALSE}
#Load all packages
library("tximport")
library("readr")
library("ggplot2")
library("DESeq2")
library("dplyr")
library("tibble")
library("sva")
library("limma")
library("ggthemr")
library("RColorBrewer")
for (package in (.packages()) ) {
 print(paste("Package", package, "version", packageVersion(package)))
}

# Set colours
cbp1 <- cbp1 <- c("#6A6A6A", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
custom_theme <- ggthemr::define_palette(
  swatch = cbp1,
  gradient = c(lower = "#6A6A6A", upper = "#56B4E9"))
ggthemr::ggthemr(custom_theme)
```

## Introduction

# <b>Questions:</b>
  1. Are all nAChR subunits expressed in the brain of A. mellifera queens and workers?
  2. Are there differences in nAChRs expression between queens and workers of A. mellifera?

Datasets used:
<br><b>worker brains</b>
<br>-Jasper et al. 2015
<br>-Christen et al. 2018
<br><b>queen brains</b>
<br>-Liberti et al. 2015
<br>-Manfredini et al. 2015

## Load data and prepare it fo analysis
- The data was loaded using tximport, metadata (.csv) was composed and imported

```{r import_files_&_metadata, include = FALSE}
### Read kallisto output and metadata ###

# Set sample identifiers using the names of the kallisto output directories
sample_ids <- dir(file.path("kallisto_results_trimmed/cast"))

# Read metadata table
samples_table <- read.csv(
  file.path("kallisto_results_trimmed/metadata", "cast.csv"),
     header = TRUE,
       stringsAsFactors = TRUE)

# Read .h5 files in
filenames <- file.path(
  "kallisto_results_trimmed/control",
  samples_table$sample,
  "abundance.h5")

# Assign names to files
names(filenames) <- samples_table$sample

# check that all files were assigned 
all(file.exists(filenames)) 
```
- I used biomaRt to associate transcripts to genes (from ensmbl metazoa)
- I imported Kallisto output
```{r biomaRt_tximport, include = FALSE}
#### Associate transcripts to genes ####
biomaRt::listMarts(host = "metazoa.ensembl.org")
# biomaRt::listMarts(host = "eg40-metazoa.ensembl.org") # version 40
# biomaRt::listEnsemblArchives()
metazoa_mart <- biomaRt::useMart(
  biomart = "metazoa_mart",
  dataset = "amellifera_eg_gene",
  host = "metazoa.ensembl.org")

transcript_to_gene <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
  mart = metazoa_mart)

# Remove all biomaRt cached files
#biomaRt::biomartCacheInfo()
biomaRt::biomartCacheClear()
#biomaRt::listAttributes(metazoa_mart)

# Import kallisto quantifications
txi_kallisto <- tximport::tximport(filenames,
  type = "kallisto",
  tx2gene = transcript_to_gene,
  txOut = FALSE)

```
- Before using DESeq2 I imported the data to SVA 
- I imported full and null model (.csv)
- I extracted one surrogate variable
```{r SVA, message=FALSE, warning=FALSE}
abundances <- txi_kallisto["abundance"]
sva_data <- as.data.frame(abundances)

# Extract only expressed genes
# (more than 10 reads in at least two samples for each gene)
dim(sva_data)
filtered <- apply(sva_data,1,function(x) length(x[x>10])>=4)
df_filtered <- sva_data[filtered,]
dim(df_filtered)
```
```{r insert_models, include=FALSE}
#   Read in full_model matrix (~cast)
full <- read.csv(
  file.path("kallisto_results_trimmed/metadata", "full_model_cast.csv"),
  header = TRUE,
  row.names = 1,
  stringsAsFactors = TRUE)
 
#   Read in null model matrix (~1)
null <- read.csv(
  file.path("kallisto_results_trimmed/metadata", "null_model_cast.csv"),
  header = TRUE,
  row.names = 1,
  stringsAsFactors = TRUE)
 
# Both as matrix
mod_null <- as.matrix(null)
mod_full <- as.matrix(full)

# Delete all the "abundance." from the names
colnames(df_filtered) <- sub("abundance.", "", colnames(df_filtered))

# Change to matrix
sva.expression <- as.matrix(df_filtered)
```
- I found a single SV 
- I plot SV against 'study'
- I ran a linnear model to check if study correlates with sv
```{r variables, message=FALSE, warning=FALSE}
# Find estimated number of surrogate variables
n.sv <- sva::num.sv(sva.expression, mod_full, method = "be", B=30)

# Running final SVA
#sva(expression dataset,model,null-model, number of variables of interest)
sva <- sva::sva(sva.expression, mod_full, mod_null, n.sv = n.sv)

# Transform for ggplot2
sv1 <- data.frame(sva$sv[,1])
rownames(sv1) <- rownames(mod_null)
study <- data.frame(samples_table[,3])
rownames(study) <- samples_table[,1]
df_merged <- merge(sv1, study, by = 0)

# Boxplot
ggplot(df_merged, aes(x = samples_table...3., y = sva.sv...1.)) +
  geom_boxplot(aes(colour = samples_table...3.),fill="white") + 
  geom_jitter(aes(colour = samples_table...3.)) 
  

# Checking if corelates with the study
summary(lm(sva$sv[,1] ~ study$samples_table...3.))
```
- I ran DESeq2
```{r DESeq2_data, message= FALSE}
# Add calculated sv to metadata table 
sv1$sample <- rownames(sv1)
merged_samples_table <- merge(samples_table, sv1, by="sample")

# Set variables for the model
cast <- as.character(unique(merged_samples_table$condition))
study <- as.character(unique(merged_samples_table$study))
covariant <- as.numeric(unique(merged_samples_table$sva.sv...1.))

# Create a DESeq object
dds <- DESeq2::DESeqDataSetFromTximport(txi_kallisto,
                                        colData = merged_samples_table,
                                        design = ~ condition + sva.sv...1.)
# Select expressed genes
dim(dds)
keep <- rowSums(DESeq2::counts(dds) >= 10) >= 4
dds <- dds[keep,]
dim(dds)

# Run differential expression analyses
# Run DESeq2
cast_dds1 <- DESeq2::DESeq(dds)
DESeq2::resultsNames(cast_dds1)

# Make vectors with selected genes of interest
gene_id <- c("GB40923", "GB53053", "GB42644", 
             "GB53055", "GB47845", "GB43416", 
             "GB42850", "GB43275", "GB53427", 
             "GB53428", "GB50159")
names_AChRs <- c("alpha8","alpha7","alpha2","beta1","alpha3",
                 "alpha6","alpha1","alpha4","alpha9","beta2","alpha5")

# Model results: cast
res_cast <- results(cast_dds1, contrast = c("condition", "queen", "worker"))

# Select genes of interest
results <- res_cast[gene_id,]
rownames(results) <- names_AChRs
results
```

## Data visualisation 
(using VST and limma)

```{r vst_heatmap, message = FALSE}
# Calculate VST
vst <- vst(dds, blind=FALSE)

# Remove batch effect from the vst data
subset.vst <- assay(vst)
subset.vst <- vst[gene_id,]
rownames(subset.vst) <- names_AChRs
subset.vst <- assay(subset.vst)
subset.vst <- limma::removeBatchEffect(subset.vst, covariates = vst$sva.sv...1., design = full)
subset.vst <- as.data.frame(subset.vst)

# Set row names as a column
subset.vst <- subset.vst[ , order(names(subset.vst))]
plot.vst <- subset.vst %>% tibble::rownames_to_column("AChRs")
plot.vst<- data.frame(plot.vst[1], stack(plot.vst[2:19]))
col_names_plot <- c("nAChRs", "vst", "cast")
colnames(plot.vst) <- col_names_plot

# Plot heatmap
ggplot(plot.vst, aes(x=cast, y=nAChRs, fill=vst)) + 
  geom_tile(color="white") +
  scale_fill_gradient(low = "wheat", high = "turquoise4") + 
  theme_minimal() +
  theme(axis.text.x=element_text(angle=65, hjust=1)) +
  xlab("nurse tissues") +
  ylab("nAChRs") +
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_text(colour="gray20", size=10),
        axis.text.y = element_text(colour="gray20", size=10))
```


```{r vst_boxplot, message = FALSE}
# Plot boxplot
bxp.vst <- plot.vst %>% mutate_all(~gsub("_1|_2|_3|_4|_5|_6", "", .)) %>% mutate_all(~gsub("_c|_j|_m|_l|_h", "", .))
bxp.vst[,2] <- as.numeric(bxp.vst$vst)
#
ggplot(data=bxp.vst, aes(x=cast, y=vst, colour=cast)) + facet_grid(~ nAChRs) +
  geom_boxplot(fill="white") + ylab("vst") + xlab("casts") +
  geom_jitter(size=1.2, pch=16, aes(colour=cast), alpha=0.45) + 
  scale_y_sqrt() + 
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_blank(),
        axis.text.y = element_text(colour="gray20", size=10)) #scale_color_brewer(palette="Dark2")
```


```{r vst_pca, message = FALSE}
# Plot PCA

# flip data frame
flip.vst <- t(subset.vst)
# perform pca
pca_res <- prcomp(flip.vst, scale.=FALSE)
summary(pca_res)
# transform data frame for ggplot2
# call out another dimention from pca output
res_out <- as.data.frame(pca_res$x)
# rownames to column to get another variable to manipulate colours in ggplot2
res_out_pca <- res_out %>% rownames_to_column("cast")
# add study as a column
res_out_pca <- res_out_pca %>% mutate(study = cast) %>% mutate_all(~gsub("_1|_2|_3|_4|_5|_6", "", .)) 
#%>% mutate_all(~gsub("queen_|worker_", "", .))
# remove redundant digits and patterns
res_out_pca$cast <- stringr::str_remove(res_out_pca$cast, "_c|_j|_m|_l|_h")
res_out_pca$study <- stringr::str_remove(res_out_pca$study, "queen_|worker_")
#change to numeric 
sapply(res_out_pca, mode)
res_out_pca <- transform(res_out_pca, PC1 = as.numeric(PC1), 
                         PC2 = as.numeric(PC2), PC3 = as.numeric(PC3), PC4 = as.numeric(PC4))
ggplot(res_out_pca, aes(x=PC1, y=PC2, colour=cast)) + 
  geom_point(aes(pch=study)) + theme_classic() +
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_text(colour="gray20", size=10),
        axis.text.y = element_text(colour="gray20", size=10))
ggplot(res_out_pca, aes(x=PC1, y=PC3, colour=cast)) + 
  geom_point(aes(pch=study)) + theme_classic() +
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_text(colour="gray20", size=10),
        axis.text.y = element_text(colour="gray20", size=10))
ggplot(res_out_pca, aes(x=PC2, y=PC3, colour=cast)) + 
  geom_point(aes(pch=study)) + theme_classic() +
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_text(colour="gray20", size=10),
        axis.text.y = element_text(colour="gray20", size=10))
```

# 1.Extracting abundances from the data and making a data frame for all samples. 
# 2.Extracting genes of interest.

```{r ratios, include = FALSE}
# Extract abundances from the imported list and make it a data frame
abundances <- txi_kallisto["abundance"]
abundances <- as.data.frame(abundances)
# Extract genes of interest
gene_list <- abundances[gene_id,]
row.names(gene_list) <- names_AChRs

```
# Normalising TPMs

```{r normalisation, message = FALSE}
# Calculate sums of each collumn
col_sum <- gene_list %>% dplyr::summarize_if(is.numeric, sum)
col_sum

ratios <- mapply('/', gene_list, col_sum)
row.names(ratios) <- names_AChRs
ratios


```


# Plotting the data

```{r visualisation, message = FALSE}
# Transform data frame
ratios <- as.data.frame(ratios)
ratios_transf <- ratios[ , order(names(ratios))]
plot.list <- ratios_transf %>% tibble::rownames_to_column("AChRs")
ncol(plot.list)
plot.list <- data.frame(plot.list[1], stack(plot.list[2:18]))
col_names_plot <- c("nAChRs", "tpm", "castes")
colnames(plot.list) <- col_names_plot
# create a df of samples which I lated add up to the plot.list
sample <- plot.list$castes
#Remove repeating pattern in castes 
plot.list <- plot.list %>% mutate_all(~gsub("abundance.|_1|_2|_3|_4|_5|_6|_j|_c|_m|_l", "", .))
# Make TPMs numeric
plot.list$tpm <- as.numeric(plot.list$tpm)
# Add sample 
plot.list$sample <- sample

custom_colors <- colorRampPalette(cbp1)(12)
ggthemr::set_swatch(custom_colors)

# Plot data
ggplot(plot.list, aes(x=castes, y=tpm, fill=nAChRs)) +
geom_bar(width = 0.7, stat = "identity", position = "fill") +
  xlab("castes") +
  ylab("normalised TPM") +
  theme(axis.title.x = element_text(colour="gray20", size=12),
        axis.title.y = element_text(colour="gray20", size=12),
        axis.text.x = element_text(colour="gray20", size=10, angle=65, hjust=1),
        axis.text.y = element_text(colour="gray20", size=10)) + scale_fill_brewer(palette="Paired")


```
# I use logistic regression to calculate p values for this model

```{r logit, message = FALSE}

# Trying to replace alpha8 with aalpha8 (a8 as a reference, logit takes them alphabetically, didn't know how to do this otherwise)
plot.list <- plot.list %>% mutate(nAChRs = ifelse(as.character(nAChRs) == "alpha8", "aalpha8", as.character(nAChRs)))

# Run model
logistic <- glm(tpm ~ nAChRs*castes , data = plot.list, family = quasibinomial(link="logit"))
summary(logistic)

# Turn glm results into a tidy tibble
broom::tidy(logistic)
# Look at the diagnostics
broom::glance(logistic)
# Return a tibble with information about data points
broom::augment(logistic)

```







